api/api/routers/mcp/_mcp_observability_middleware.py

```diff
@@ -0,0 +1,341 @@
+                    yield chunk_bytes
+
+            except Exception as e:
+                logger.error(
+                    "Error during streaming response iteration",
+                    extra={"error": str(e)},
+                )
+                raise
+            finally:
+                # Process complete response after streaming is done
+                if body_parts:
+                    complete_body = b"".join(body_parts)
+                    try:
+                        on_complete_callback(complete_body)
+                    except Exception as e:
+                        logger.warning(
+                            "Error in streaming response completion callback",
+                            extra={"error": str(e)},
+                        )
+
+        return StreamingResponse(
+            observing_body_iterator(),
+            status_code=original_response.status_code,
+            headers=original_response.headers,
+            media_type=original_response.media_type,
+        )
+
+    def _parse_tool_call_response(self, response_body: bytes, request_id: str) -> str:
+        try:
+            response_str = response_body.decode("utf-8")
+            response_json_str = extract_json_str(response_str)
+
+            response_data = json.loads(response_json_str)
+            return response_data["result"]["content"][0]["text"]
+
+        except (json.JSONDecodeError, UnicodeDecodeError, KeyError, IndexError, ValueError) as e:
+            logger.error(
+                "Failed to parse MCP response JSON",
+                extra={
+                    "error": str(e),
+                    "response_body_length": len(response_body),
+                    "request_id": request_id,
+                },
+            )
+            return str(e)
+
+    async def _run_observer_agent_background(self, observer_data: ObserverAgentData):
+        """Run observer agent in background for analysis"""
+        try:
+
+            async def run_observer_agent() -> None:
+                try:
+                    await mcp_tool_call_observer_agent(
+                        tool_name=observer_data.tool_name,
+                        previous_tool_calls=observer_data.previous_tool_calls,
+                        tool_arguments=observer_data.tool_arguments,
+                        tool_result=observer_data.tool_result,
+                        duration_seconds=observer_data.duration_seconds,
+                        user_agent=observer_data.user_agent,
+                        mcp_session_id=observer_data.mcp_session_id,
+                        organization_name=observer_data.organization_name,
+                    )
+                except Exception as e:
+                    logger.exception(
+                        "Observer agent execution failed",
+                        exc_info=e,
+                    )
+
+            # Run in the background
+            add_background_task(run_observer_agent())
+        except Exception as e:
```

Comment:
no need to wrap here. why not just expose `run_observer_agent(observer_data:` and call ` add_background_task(run_observer_agent(observer_data))` directly. 

add_background_task is sufficiently shielded

Reply:
corrected